API.md

```diff
@@ -0,0 +1,319 @@
+- the way in which plugins communicate user intent to OpenSCD core.
+- the way in which OpenSCD sets CSS fonts and colors for plugins
+
+## Communicating data to plugins
+
+OpenSCD core communicates the data necessary for editing SCL documents by setting the following [properties](https://developer.mozilla.org/en-US/docs/Glossary/Property/JavaScript) on the plugin's [DOM Element](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement):
+
+
+```typescript=
+export default class Plugin extends LitElement {
+  docs: Record<string, XMLDocument> = {};
+  doc?: XMLDocument;
+  docName?: string;
+  editable: string[] = [
+    'cid',
+    'icd',
+    'iid',
+    'scd',
+    'sed',
+    'ssd',
+  ];
+  history: LogEntry[];
+  editCount: number = 0;
+  plugins: { menu: Plugin[], editor: Plugin[] }[];
+  locale: string = 'en';
+}
+
+/** Helper types exported by OpenSCD core **/
+
+type LogEntry = {
+  redo: Edit; // `Edit` defined below
+  undo: Edit;
+  title?: string
+}
+
+type Plugin = {
+  name: string;
+  translations?: Record<string, string>;
+  src: string;
+  icon: string;
+  requireDoc?: boolean; // disable plugin if no doc is opened
+}
+```
+
+### `docs`
+
+`docs` is set to an object mapping `string` keys (document names) to `XMLDocument` values.
+
+### `docName`
+The name of the `XMLDocument` currently being edited.
+
+### `doc`
+The `XMLDocument` currently being edited.
+
+### `editable`
+Filename extensions of user-editable documents.
+
+### `history`
+History of edits done to `doc`.
+
+### `editCount`
+
+Index of the current log entry in the `history`. Is incremented with every new edit and updated on undo/redo. 
+
+### `plugins`
+
+Arrays of `Plugin` objects describing the currently loaded `menu` and `editor` plugins, respectively.
+
+### `locale`
+
+Selected language (IETF language tag).
+
+## Communicating user intent to OpenSCD core
+
+Plugins communicate user intent to OpenSCD core by dispatching the following [custom events](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent):
+
+### `EditEventV2`
+
+The **edit event** allows a plugin to describe the changes it wants to make to the current `doc`.
+
+```typescript=
+export type EditDetailV2<E extends Edit = Edit> = {
+  edit: E;
+  title?: string;
+  squash?: boolean;
+}
+
+export type EditEventV2<E extends Edit = Edit> = CustomEvent<EditDetailV2<E>>;
+
+export type EditEventOptions = {
+  title?: string;
+  squash?: boolean;
+}
+
+export function newEditEventV2<E extends Edit>(edit: E, options: EditEventOptions): EditEventV2<E> {
+  return new CustomEvent<E>('oscd-edit-v2', {
+    composed: true,
+    bubbles: true,
+    detail: {...options, edit},
+  });
+}
+
+declare global {
+  interface ElementEventMap {
+    ['oscd-edit-v2']: EditEventV2;
+  }
+}
+```
+
+It's `title` property is a human-readable description of the edit, and `squash` is a boolean indicating whether the edit should be merged with the previous edit in the history.
```

Comment:
```suggestion
It's `title` property is a human-readable description of the edit (for display in editing history), and `squash` is a boolean indicating whether the edit should be merged with the previous edit in the history.
```

Do I understand the purpose correctly that the title is mainly for display in editing history?

Reply:
Yes. At least that is how I did understand this. It makes sense mostly with complex edits like the importing of an IED, where - if you want to display a history to the user - it would be beneficial to have a better description of the complex edit in the hostory.

Reply:
No other app that I know of even displays an "edit history". Microsoft Word, e.g., only shows "undo" and "redo" buttons, respectively, and doesn't tell you what it is that would be undone or redone.

If we were to remove this "title" feature and the "edit history" view, we'd have a simpler API and avoid the problem of history entries not being translatable once they have been recorded. I am in favour of this approach.

What do you think, @danyill and @JakobVogelsang ?

Reply:
I am torn about that. I am using it a lot at the moment, BUT if I am being honest, that is more for debugging while plugin development than anything else. Other users have reported a different story.

That said, I still would provide the history, because other plugin might want to better display the history or do something like writing to the `History` item. I would also still have the `sqash` key now in order to have correct redo states, which is a bit tricky sometimes at the moment.

Reply:
Of course, the `squash` flag is needed for undo and redo to not infuriate end users. Whether the `title` is really needed is still an open question, though. For `Hitem`s a hand-written piece of rich text by the end user that describes a whole bunch of related edits (much like a `git` commit message) seems more sensible to me. I'd love to have more feedback from actual users (i.e. engineers) on this.

Reply:
> No other app that I know of even displays an "edit history". Microsoft Word, e.g., only shows "undo" and "redo" buttons, respectively, and doesn't tell you what it is that would be undone or redone.

Oh but it does! However over the undo and redo buttons some time ;-) I remember it from before the ribbon days when it was suffixed to the words "Undo" and "Redo" in the drop-down menu.

Reply:
> Whether the title is really needed is still an open question, though.

I agree it's an open question but Microsoft Word and others do have a similar feature.
They also allow you to multiselect a number of changes down to a certain level to undo them which I do think is rather nice (and not a feature I use very often):

![image](https://github.com/user-attachments/assets/a4852b2d-a7f5-4550-873b-a3bb4070db93)

Interestingly the same menu is not maintained for the "Redo" function.

Reply:
My feeling is, "If in doubt we should add it". I like Jakob's point about it being useful for development and I also think users often want an assurance on a step of actions recently taken (particular if they've been clicking a lot of buttons with not directly visible consequences). Being able to see a drop down even just to see "what I did recently" has some utility.

Reply:
> > No other app that I know of even displays an "edit history". Microsoft Word, e.g., only shows "undo" and "redo" buttons, respectively, and doesn't tell you what it is that would be undone or redone.
> 
> Oh but it does! However over the undo and redo buttons some time ;-) I remember it from before the ribbon days when it was suffixed to the words "Undo" and "Redo" in the drop-down menu.

Thanks for the lesson! It really shows that I don't use Microsoft software day to day.

> My feeling is, "If in doubt we should add it".

Generally, my feeling is "If in doubt, leave it out!", but in this particular case I can see the usefulness. I also always wanted to have the feature where if you click on a non-current edit, you jump to the state of the document just after that edit. If you click on the current edit, you "undo". That way any place in the edit history is at most a double-click away. Would be fairly trivial to implement in core, but I haven't found the time for that nice-to-have yet.